tests/cpp/objective/test_aft_obj.cc

```diff
@@ -0,0 +1,174 @@
+// Then check prediction against the reference values
+static inline void CheckGPairOverGridPoints(
+                      ObjFunction* obj,
+                      bst_float true_label_lower_bound,
+                      bst_float true_label_upper_bound,
+                      const std::string& dist_type,
+                      const std::vector<bst_float>& expected_grad,
+                      const std::vector<bst_float>& expected_hess,
+                      float ftol = 1e-4f) {
+  const int num_point = 20;
+  const double log_y_low = 1.0;
+  const double log_y_high = 15.0;
+  
+  obj->Configure({ {"aft_loss_distribution", dist_type},
+                   {"aft_loss_distribution_scale", "1"} });
+
+  MetaInfo info;
+  info.num_row_ = num_point;
+  info.labels_lower_bound_.HostVector()
+    = std::vector<bst_float>(num_point, true_label_lower_bound);
+  info.labels_upper_bound_.HostVector()
+    = std::vector<bst_float>(num_point, true_label_upper_bound);
+  info.weights_.HostVector() = std::vector<bst_float>();
+  std::vector<bst_float> preds(num_point);
+  for (int i = 0; i < num_point; ++i) {
+    preds[i] = std::log(std::pow(2.0, i * (log_y_high - log_y_low) / (num_point - 1) + log_y_low));
+  }
+
+  HostDeviceVector<GradientPair> out_gpair;
+  obj->GetGradient(HostDeviceVector<bst_float>(preds), info, 1, &out_gpair);
+  const auto& gpair = out_gpair.HostVector();
+  CHECK_EQ(num_point, expected_grad.size());
+  CHECK_EQ(num_point, expected_hess.size());
+  for (int i = 0; i < num_point; ++i) {
+    EXPECT_NEAR(gpair[i].GetGrad(), expected_grad[i], ftol);
+    EXPECT_NEAR(gpair[i].GetHess(), expected_hess[i], ftol);
+  }
+}
+
+TEST(Objective, AFTObjGPairUncensoredLabels) {
+  auto lparam = CreateEmptyGenericParam(-1);  // currently AFT objective is CPU only
+  std::unique_ptr<ObjFunction> obj(ObjFunction::Create("aft:survival", &lparam));
+
+  CheckGPairOverGridPoints(obj.get(), 100.0f, 100.0f, "normal",
+    { -3.9120f, -3.4013f, -2.8905f, -2.3798f, -1.8691f, -1.3583f, -0.8476f, -0.3368f, 0.1739f,
+      0.6846f, 1.1954f, 1.7061f, 2.2169f, 2.7276f, 3.2383f, 3.7491f, 4.2598f, 4.7706f, 5.2813f,
+      5.7920f },
+    { 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f,
+      1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f, 1.0000f });
+  CheckGPairOverGridPoints(obj.get(), 100.0f, 100.0f, "logistic",
+    { -0.9608f, -0.9355f, -0.8948f, -0.8305f, -0.7327f, -0.5910f, -0.4001f, -0.1668f, 0.0867f,
+      0.3295f, 0.5354f, 0.6927f, 0.8035f, 0.8773f, 0.9245f, 0.9540f, 0.9721f, 0.9832f, 0.9899f,
+      0.9939f },
+    { 0.0384f, 0.0624f, 0.0997f, 0.1551f, 0.2316f, 0.3254f, 0.4200f, 0.4861f, 0.4962f, 0.4457f,
+      0.3567f, 0.2601f, 0.1772f, 0.1152f, 0.0726f, 0.0449f, 0.0275f, 0.0167f, 0.0101f, 0.0061f });
+  CheckGPairOverGridPoints(obj.get(), 100.0f, 100.0f, "extreme",
+    { -0.0000f, -29.0026f, -17.0031f, -9.8028f, -5.4822f, -2.8897f, -1.3340f, -0.4005f, 0.1596f,
```

Comment:
@avinashbarnwal FYI, I applied the regularization scheme to the uncensored case as well, and now I'm getting a zero gradient here, where previously we'd get something like -50.0. I'm still looking at ways to avoid INF and NAN without strange behavior like this. For this example, clamping the gradient to a reasonable quantity like -30.0 would be a lot better than giving 0.0.

Reply:
Thanks

Reply:
I’ve merged this PR for now. I’ll file a follow up PR to make AFT more robust in edge cases like this.